src/Plugins/ComplexCore/src/ComplexCore/Filters/InitializeImageGeomCellData.cpp

```diff
@@ -0,0 +1,369 @@
+  }
+  case to_underlying(InitializeImageGeomCellData::InitType::Random): {
+    return InitializeImageGeomCellData::InitType::Random;
+  }
+  case to_underlying(InitializeImageGeomCellData::InitType::RandomWithRange): {
+    return InitializeImageGeomCellData::InitType::RandomWithRange;
+  }
+  default: {
+    throw std::runtime_error("InitializeImageGeomCellData: Invalid value for InitType");
+  }
+  }
+}
+
+struct CheckInitializationFunctor
+{
+  template <class T>
+  std::optional<Error> operator()(const IDataArray& dataArray, InitializeImageGeomCellData::InitType initType, float64 initValue, const std::pair<float64, float64>& initRange)
+  {
+    std::string arrayName = dataArray.getName();
+
+    if(initType == InitializeImageGeomCellData::InitType::Manual)
+    {
+      if(initValue < static_cast<double>(std::numeric_limits<T>().lowest()) || initValue > static_cast<double>(std::numeric_limits<T>().max()))
+      {
+        return Error{-4000, fmt::format("{}: The initialization value could not be converted. The valid range is {} to {}", arrayName, std::numeric_limits<T>::min(), std::numeric_limits<T>::max())};
+      }
+    }
+    else if(initType == InitializeImageGeomCellData::InitType::RandomWithRange)
+    {
+      float64 min = initRange.first;
+      float64 max = initRange.second;
+      if(min > max)
+      {
+        return Error{-5550, fmt::format("{}: Invalid initialization range.  Minimum value is larger than maximum value.", arrayName)};
+      }
+      if(min < static_cast<double>(std::numeric_limits<T>().lowest()) || max > static_cast<double>(std::numeric_limits<T>().max()))
+      {
+        return Error{-4001, fmt::format("{}: The initialization range can only be from {} to {}", arrayName, std::numeric_limits<T>::min(), std::numeric_limits<T>::max())};
+      }
+      if(min == max)
+      {
+        return Error{-4002, fmt::format("{}: The initialization range must have differing values", arrayName)};
+      }
+    }
+
+    return {};
+  }
+};
+
+template <class T>
+auto CreateRandomGenerator(T rangeMin, T rangeMax, uint64 seed)
+{
+  std::random_device randomDevice;           // Will be used to obtain a seed for the random number engine
+  std::mt19937_64 generator(randomDevice()); // Standard mersenne_twister_engine seeded with rd()
+  generator.seed(seed);
+
+  if constexpr(std::is_integral_v<T>)
+  {
+    std::uniform_int_distribution<> distribution(rangeMin, rangeMax);
+    return std::make_pair(distribution, generator);
+  }
+  else if constexpr(std::is_floating_point_v<T>)
+  {
+    std::uniform_real_distribution<T> distribution(rangeMin, rangeMax);
+    return std::make_pair(distribution, generator);
+  }
+}
+
+struct InitializeArrayFunctor
+{
+  template <class T>
+  void operator()(IDataArray& dataArray, const std::array<usize, 3>& dims, uint64 xMin, uint64 xMax, uint64 yMin, uint64 yMax, uint64 zMin, uint64 zMax, InitializeImageGeomCellData::InitType initType,
```

Comment:
> Since this is templated, can you use DataArray? That also slightly simplifies line 123 so you can use dataArray.getDataStoreRef() to receive the AbstractDataStore reference.

This was the original Initialize Data filter that was renamed so I didn't really work on this one, hence the extensive repeated use of switches in the test for it, but I don't think this is possible with the API call to the ExecuteNeighborFunction. That being said I can turn this functor into a class and capture the typed DataArray as a member variable from the constructor via the RunTemplateClass API function if that is preferable.

Reply:
This is not possible without a switch to grab the array from the dataStrucutre or passing the whole datastructure by ref to the functor